pkg/controller/cainjector/controller.go

```diff
@@ -0,0 +1,211 @@
+// of inject target.  For instance, an implementation might create a InjectTarget
+// containing an empty MutatingWebhookConfiguration.  The underlying API object can
+// be populated (via AsObject) using client.Client#Get, and then CAs can be injected with
+// Injectables (representing the various individual webhooks in the config) retrieved with
+// Services.
+type CertInjector interface {
+	// NewTarget creates a new InjectTarget containing an empty underlying object.
+	NewTarget() InjectTarget
+}
+
+// genericInjectReconciler is a reconciler that knows how to check if a given object is
+// marked as requiring a CA, chase down the corresponding Service, Certificate, Secret, and
+// inject that into the object.
+type genericInjectReconciler struct {
+	// injector is responsible for the logic of actually setting a CA -- it's the component
+	// that contains type-specific logic.
+	injector CertInjector
+
+	log logr.Logger
+	client.Client
+
+	resourceName string // just used for logging
+}
+
+// InjectClient provides a client for this reconciler.
+func (r *genericInjectReconciler) InjectClient(c client.Client) error {
+	r.Client = c
+	return nil
+}
+
+// splitNamespacedName turns the string form of a namespaced name
+// (<namespace>/<name>) back into a types.NamespacedName.
+func splitNamespacedName(nameStr string) types.NamespacedName {
+	splitPoint := strings.IndexRune(nameStr, types.Separator)
+	if splitPoint == -1 {
+		return types.NamespacedName{Name: nameStr}
+	}
+	return types.NamespacedName{Namespace: nameStr[:splitPoint], Name: nameStr[splitPoint+1:]}
+}
+
+// Reconcile attempts to ensure that a particular object has all the CAs injected that
+// it has requested.
+func (r *genericInjectReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.log.WithValues(r.resourceName, req.NamespacedName)
+
+	// fetch the target object
+	target := r.injector.NewTarget()
+	if err := r.Client.Get(ctx, req.NamespacedName, target.AsObject()); err != nil {
+		log.Error(err, "unable to fetch target object to inject into")
+		return ctrl.Result{}, err
+	}
+
+	// ensure that it wants injection
+	metaObj, err := meta.Accessor(target.AsObject())
+	if err != nil {
+		log.Error(err, "unable to get metadata for object")
+		return ctrl.Result{}, err
+	}
+	certNameRaw := metaObj.GetAnnotations()[WantInjectAnnotation]
+	if certNameRaw == "" {
+		log.V(1).Info("object does not want CA injection, skipping")
+		return ctrl.Result{}, nil
+	}
+
+	certName := splitNamespacedName(certNameRaw)
+	log = log.WithValues("certificate", certName)
+	if certName.Namespace == "" {
+		log.Error(nil, "invalid certificate name")
+		// don't return an error, requeuing won't help till this is changed
+		return ctrl.Result{}, nil
+	}
+
+	var cert certmanager.Certificate
+	if err := r.Client.Get(ctx, certName, &cert); err != nil {
+		log.Error(err, "unable to fetch associated certificate")
+		return ctrl.Result{}, err
```

Comment:
What's the rate limiter like with the default controller-runtime queue? If the Certificate resource doesn't exist, we'll start looping here. I know the default limiter in client-go has a base of 2ms and a cap of 1s, which results in probably more traffic than is really necessary 😄

Reply:
I can add a "does not exist" check -- I tried to just not return an error most cases where that was the case.